### PR TITLE
if_then_else support via "conditioned" data frames

### PR DESCRIPTION
**Pull request by ramiromagno**
- Adds a new S3 class (cnd_df) for represented conditioned data frames, i.e. data frames that carry metadata about what records should be used for derivations

- Adds support for basic pretty printing of cnd_df objects

- Adds a user-facing function for creating such cnd_df objects: `condition_by`

- Adds experimental "mutate"-version function for these conditioned data frames: `derive_by_condition()`

Thank you for your Pull Request! We have developed this task checklist from the
[Development Process
Guide](https://pharmaverse.github.io/sdtm.oak/CONTRIBUTING.html)
to help with the final steps of the process. Completing the below tasks helps to
ensure our reviewers can maximize their time on your code as well as making sure
the oak codebase remains robust and consistent.

Please check off each taskbox as an acknowledgment that you completed the task
or check off that it is not relevant to your Pull Request. This checklist is
part of the Github Action workflows and the Pull Request will not be merged into
the `devel` branch until you have checked off each task.

- [x] Place Closes #<insert_issue_number> into the beginning of your Pull z
Request Title (Use Edit button in top-right if you need to update)
- [x] Code is formatted according to the 
[tidyverse style guide](https://style.tidyverse.org/). Run 
`styler::style_file()` to style R and Rmd files
- [x] Updated relevant unit tests or have written new unit tests, which should
consider realistic data scenarios and edge cases, e.g. empty datasets, errors,
boundary cases etc. - See 
[Unit Test Guide](https://pharmaverse.github.io/admiraldev/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
- [x] If you removed/replaced any function and/or function parameters, did you 
fully follow the 
[deprecation guidance](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#deprecation)?
- [x] Update to all relevant roxygen headers and examples, including keywords
and families. Refer to the 
[categorization of functions](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#categorization-of-functions) to tag appropriate keyword/family.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the
`NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md` if the changes pertain to a user-facing function (i.e. it
has an `@export` tag) or documentation aimed at users (rather than developers)
- [x] Build oak site `pkgdown::build_site()` and check that all affected
examples are displayed correctly and that all new functions occur on the "[Reference](https://pharmaverse.github.io/admiral/reference/index.html)" page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue in the Development Section on the right hand side.
- [x] Address all merge conflicts and resolve appropriately
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

### CODE REVIEW

**Comment by ramiromagno** [comment]
Still work in progress.

**Comment by rammprasad** [comment]
@ramiromagno - 

A couple of items
1. the condition_by is not working when applied at the target variable
2. Also, can you show me how to apply condition_by when involving both source and target?

Here is the sample code.

```R

study_ct <- read.csv(system.file("cm_domain/cm_sdtm_oak_ct.csv",
                                 package = "sdtm.oak"))

# Read in raw data

cm_raw <-  read.csv(system.file("cm_domain/cm_raw_data.csv",
                                package = "sdtm.oak")) %>%
  #derive oak_id as the row number
  dplyr::mutate(oak_id = structure(seq_len(nrow(.))),
                patient_number = PATNUM,
                raw_source = "ConMed") %>%  
  dplyr::select(oak_id_vars(), dplyr::everything())

# Create CM domain. The first step in creating CM domain is to create the topic variable

cm <-
  # Derive topic variable
  assign_no_ct(
    raw_dat = cm_raw,
    raw_var = "MDRAW",
    tgt_var = "CMTRT"
  )  |>
  # Derive CMGRPID when CMTRT == "BABY ASPIRIN"
  assign_no_ct(
    raw_dat = cm_raw,
    raw_var = "MDNUM",
    tgt_var = "CMGRPID",
    #use condition_by function and fiter when CMTRT is "BABY ASPIRIN"
   ### This condition is not working. It results in an error
    tgt_dat = condition_by(.,CMTRT == "BABY ASPIRIN"),
    id_vars = oak_id_vars()
  )

**Error message**
Error in assign_no_ct(assign_no_ct(raw_dat = cm_raw, raw_var = "MDRAW",  : 
  unused argument (assign_no_ct(raw_dat = cm_raw, raw_var = "MDRAW", tgt_var = "CMTRT"))

Can you give give an example on how to program this mapping?
  # Derive qualifier CMMODIFY -  If collected value in CMMODIFY
  # in cm_raw is different to CM domain CMTRT target variable then
  # assign the collected value to CMMODIFY in CM domain (CM.CMMODIFY)

```

**Comment by ramiromagno** [comment]
> @ramiromagno -
> 
> A couple of items
> 
>     1. the condition_by is not working when applied at the target variable
> 
>     2. Also, can you show me how to apply condition_by when involving both source and target?
> 
> 
> Here is the sample code.
> 
> ```r
> study_ct <- read.csv(system.file("cm_domain/cm_sdtm_oak_ct.csv",
>                                  package = "sdtm.oak"))
> 
> # Read in raw data
> 
> cm_raw <-  read.csv(system.file("cm_domain/cm_raw_data.csv",
>                                 package = "sdtm.oak")) %>%
>   #derive oak_id as the row number
>   dplyr::mutate(oak_id = structure(seq_len(nrow(.))),
>                 patient_number = PATNUM,
>                 raw_source = "ConMed") %>%  
>   dplyr::select(oak_id_vars(), dplyr::everything())
> 
> # Create CM domain. The first step in creating CM domain is to create the topic variable
> 
> cm <-
>   # Derive topic variable
>   assign_no_ct(
>     raw_dat = cm_raw,
>     raw_var = "MDRAW",
>     tgt_var = "CMTRT"
>   )  |>
>   # Derive CMGRPID when CMTRT == "BABY ASPIRIN"
>   assign_no_ct(
>     raw_dat = cm_raw,
>     raw_var = "MDNUM",
>     tgt_var = "CMGRPID",
>     #use condition_by function and fiter when CMTRT is "BABY ASPIRIN"
>    ### This condition is not working. It results in an error
>     tgt_dat = condition_by(.,CMTRT == "BABY ASPIRIN"),
>     id_vars = oak_id_vars()
>   )
> 
> **Error message**
> Error in assign_no_ct(assign_no_ct(raw_dat = cm_raw, raw_var = "MDRAW",  : 
>   unused argument (assign_no_ct(raw_dat = cm_raw, raw_var = "MDRAW", tgt_var = "CMTRT"))
> 
> Can you give give an example on how to program this mapping?
>   # Derive qualifier CMMODIFY -  If collected value in CMMODIFY
>   # in cm_raw is different to CM domain CMTRT target variable then
>   # assign the collected value to CMMODIFY in CM domain (CM.CMMODIFY)
> ```

Yeah, there are a couple of things at play here:

1. The placeholder `.` is for use with magrittr's pipe `%>%` and `_` for use with R native pipe `|>`.
2. The usage of the placeholder in nested calls requires the use of braces, see https://magrittr.tidyverse.org/reference/pipe.html#using-the-dot-for-secondary-purposes. Not sure we want to surface this to the user.

So best approach might be that if we want to condition on the target data set, the one that is being passed along, we should perhaps move the `condition_by()` call one level up.

Here is a set of examples that hopefully illustrate the different variations:

``` r
library(sdtm.oak)
library(magrittr)
library(dplyr, warn.conflicts = FALSE)

study_ct <- read.csv(system.file("cm_domain/cm_sdtm_oak_ct.csv", package = "sdtm.oak"))
cm_raw <-  read.csv(system.file("cm_domain/cm_raw_data.csv", package = "sdtm.oak")) %>%
  #derive oak_id as the row number
  dplyr::mutate(
    oak_id = structure(seq_len(nrow(.))),
    patient_number = PATNUM,
    raw_source = "ConMed"
  ) %>%
  dplyr::select(sdtm.oak:::oak_id_vars(), dplyr::everything())

# Derive topic variable
assign_no_ct(raw_dat = cm_raw,
             raw_var = "MDRAW",
             tgt_var = "CMTRT")  %>%
  # Derive CMGRPID when CMTRT == "BABY ASPIRIN"
  assign_no_ct(
    raw_dat = cm_raw,
    raw_var = "MDNUM",
    tgt_var = "CMGRPID",
    tgt_dat = condition_by(., CMTRT == "BABY ASPIRIN")
  )
#> Error in `admiraldev::assert_character_vector()` at sdtm.oak/R/assign.R:189:3:
#> ! `id_vars` must be a character vector but is a data frame


# Derive topic variable
assign_no_ct(raw_dat = cm_raw,
             raw_var = "MDRAW",
             tgt_var = "CMTRT")  |>
  condition_by(CMTRT == "BABY ASPIRIN") |>
  assign_no_ct(
    raw_dat = cm_raw,
    raw_var = "MDNUM",
    tgt_var = "CMGRPID",
    tgt_dat = _
  )
#> # A tibble: 14 × 5
#>    oak_id raw_source patient_number CMTRT                         CMGRPID
#>     <int> <chr>               <int> <chr>                           <int>
#>  1      1 ConMed                375 BABY ASPIRIN                        1
#>  2      2 ConMed                375 CORTISPORIN                        NA
#>  3      3 ConMed                376 ASPIRIN                            NA
#>  4      4 ConMed                377 DIPHENHYDRAMINE HCL                NA
#>  5      5 ConMed                377 PARCETEMOL                         NA
#>  6      6 ConMed                377 VOMIKIND                           NA
#>  7      7 ConMed                377 ZENFLOX OZ                         NA
#>  8      8 ConMed                378 AMITRYPTYLINE                      NA
#>  9      9 ConMed                378 BENADRYL                           NA
#> 10     10 ConMed                378 DIPHENHYDRAMINE HYDROCHLORIDE      NA
#> 11     11 ConMed                378 TETRACYCLINE                       NA
#> 12     12 ConMed                379 BENADRYL                           NA
#> 13     13 ConMed                379 SOMINEX                            NA
#> 14     14 ConMed                379 ZQUILL                             NA


# Derive topic variable
assign_no_ct(raw_dat = cm_raw,
             raw_var = "MDRAW",
             tgt_var = "CMTRT")  |>
  condition_by(CMTRT == "BABY ASPIRIN") %>%
  assign_no_ct(
    raw_dat = cm_raw,
    raw_var = "MDNUM",
    tgt_var = "CMGRPID",
    tgt_dat = .
  )
#> # A tibble: 14 × 5
#>    oak_id raw_source patient_number CMTRT                         CMGRPID
#>     <int> <chr>               <int> <chr>                           <int>
#>  1      1 ConMed                375 BABY ASPIRIN                        1
#>  2      2 ConMed                375 CORTISPORIN                        NA
#>  3      3 ConMed                376 ASPIRIN                            NA
#>  4      4 ConMed                377 DIPHENHYDRAMINE HCL                NA
#>  5      5 ConMed                377 PARCETEMOL                         NA
#>  6      6 ConMed                377 VOMIKIND                           NA
#>  7      7 ConMed                377 ZENFLOX OZ                         NA
#>  8      8 ConMed                378 AMITRYPTYLINE                      NA
#>  9      9 ConMed                378 BENADRYL                           NA
#> 10     10 ConMed                378 DIPHENHYDRAMINE HYDROCHLORIDE      NA
#> 11     11 ConMed                378 TETRACYCLINE                       NA
#> 12     12 ConMed                379 BENADRYL                           NA
#> 13     13 ConMed                379 SOMINEX                            NA
#> 14     14 ConMed                379 ZQUILL                             NA


# Derive topic variable
assign_no_ct(raw_dat = cm_raw,
             raw_var = "MDRAW",
             tgt_var = "CMTRT")  %>%
  {
    assign_no_ct(
      raw_dat = cm_raw,
      raw_var = "MDNUM",
      tgt_var = "CMGRPID",
      tgt_dat = condition_by(dat = ., CMTRT == "BABY ASPIRIN")
    )
  }
#> # A tibble: 14 × 5
#>    oak_id raw_source patient_number CMTRT                         CMGRPID
#>     <int> <chr>               <int> <chr>                           <int>
#>  1      1 ConMed                375 BABY ASPIRIN                        1
#>  2      2 ConMed                375 CORTISPORIN                        NA
#>  3      3 ConMed                376 ASPIRIN                            NA
#>  4      4 ConMed                377 DIPHENHYDRAMINE HCL                NA
#>  5      5 ConMed                377 PARCETEMOL                         NA
#>  6      6 ConMed                377 VOMIKIND                           NA
#>  7      7 ConMed                377 ZENFLOX OZ                         NA
#>  8      8 ConMed                378 AMITRYPTYLINE                      NA
#>  9      9 ConMed                378 BENADRYL                           NA
#> 10     10 ConMed                378 DIPHENHYDRAMINE HYDROCHLORIDE      NA
#> 11     11 ConMed                378 TETRACYCLINE                       NA
#> 12     12 ConMed                379 BENADRYL                           NA
#> 13     13 ConMed                379 SOMINEX                            NA
#> 14     14 ConMed                379 ZQUILL                             NA


# Derive topic variable
assign_no_ct(raw_dat = cm_raw,
             raw_var = "MDRAW",
             tgt_var = "CMTRT")  %>%
  {
    assign_no_ct(
      raw_dat = cm_raw,
      raw_var = "MDNUM",
      tgt_var = "CMGRPID",
      tgt_dat = condition_by(., CMTRT == "BABY ASPIRIN")
    )
  }
#> # A tibble: 14 × 5
#>    oak_id raw_source patient_number CMTRT                         CMGRPID
#>     <int> <chr>               <int> <chr>                           <int>
#>  1      1 ConMed                375 BABY ASPIRIN                        1
#>  2      2 ConMed                375 CORTISPORIN                        NA
#>  3      3 ConMed                376 ASPIRIN                            NA
#>  4      4 ConMed                377 DIPHENHYDRAMINE HCL                NA
#>  5      5 ConMed                377 PARCETEMOL                         NA
#>  6      6 ConMed                377 VOMIKIND                           NA
#>  7      7 ConMed                377 ZENFLOX OZ                         NA
#>  8      8 ConMed                378 AMITRYPTYLINE                      NA
#>  9      9 ConMed                378 BENADRYL                           NA
#> 10     10 ConMed                378 DIPHENHYDRAMINE HYDROCHLORIDE      NA
#> 11     11 ConMed                378 TETRACYCLINE                       NA
#> 12     12 ConMed                379 BENADRYL                           NA
#> 13     13 ConMed                379 SOMINEX                            NA
#> 14     14 ConMed                379 ZQUILL                             NA
```

<sup>Created on 2024-06-07 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Comment by rammprasad** [comment]
Thank you, @ramiromagno. I got it to work. Having it inline will make sense for the users. We will just put both options out there.

```R

cm <-
  # Derive topic variable
  assign_no_ct(
    raw_dat = cm_raw,
    raw_var = "MDRAW",
    tgt_var = "CMTRT"
  )  %>%
  # Derive CMGRPID
  {
    assign_no_ct(
    raw_dat = cm_raw,
    raw_var = "MDNUM",
    tgt_var = "CMGRPID",
    #use condition_by function and fiter when CMTRT is "BABY ASPIRIN"
    tgt_dat = condition_by(dat =., CMTRT == "BABY ASPIRIN"),
    id_vars = oak_id_vars()
  )
    } %>%
  {
  assign_no_ct(
    raw_dat = cm_raw,
    raw_var = "CMMODIFY",
    tgt_var = "CMMODIFY",
    tgt_dat =  condition_by(. , CMMODIFY != CMTRT, .env = cm_raw),
    id_vars = oak_id_vars()
  )
  }
```

A couple of follow-up questions
1. We need to be explain in the documentation why we need {} when we use the condition_by function.
2. Can we rename .env parameter to something else more meaningful? We are comparing two datasets in this case, so can we rename it as `.dat2`

```R
  {
  assign_no_ct(
    raw_dat = cm_raw,
    raw_var = "CMMODIFY",
    tgt_var = "CMMODIFY",
    tgt_dat =  condition_by( dat = . , CMMODIFY != CMTRT,  dat2 = cm_raw),
    id_vars = oak_id_vars()
  )
  }
#option 2
  {
  assign_no_ct(
    raw_dat = condition_by(dat = cm_raw , CMMODIFY != CMTRT, dat2= .),
    raw_var = "CMMODIFY",
    tgt_var = "CMMODIFY",
    tgt_dat =  .,
    id_vars = oak_id_vars()
  )
  }

**Comment by ramiromagno** [comment]
Hi @rammprasad:

1. The reason why we need braces `{...}` is explained here: https://magrittr.tidyverse.org/reference/pipe.html#using-the-dot-for-secondary-purposes. We could get rid of this requirement if `tgt_dat` was the first parameter, and I think we should, as it would simplify the overall piping syntax, making it easier on the user.

2. Renaming `.env` to `.dat2` (I'm adding the dot) is fine but `.env` is more aligned with what is customary in the tidyverse ecosystem for scopes where to look for variables, be it an actual environment (env), or a data frame, tibble or simply a list.

File: R/cnd_df.R [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+#
+# - `new_cnd_df()`: Create a "conditioned data set" (class cnd_df).
+# - `is_cnd_df()`: Assess whether the argument is a `cnd_df` data frame.
+# - `get_cnd_df_cnd()`: Extract the attribute "cnd" from a `cnd_df` data frame.
+# - `get_cnd_df_cnd_sum()`: Extract the attribute "cnd_sum" from a `cnd_df` data frame.
+# - `rm_cnd_df()`: De-class a cnd_df data frame.
+# - `tbl_sum.cnd_df()`: Provide a tibble header print method for `cnd_df` tibbles.
+# - `ctl_new_rowid_pillar.cnd_df()`: A print method for the row ids cnd_df` tibbles.
+# - `eval_conditions()`: Find which rows match a set of conditions.
+# - `condition_by()`: Create a conditioned data frame (user facing).
```
Comment (rammprasad) [inline R/cnd_df.R:23]: Shall we change condition_by() to `condition_add`. Actually, we are adding conditions metadata to the dataframe. This also aligns with {admiral} where we have filter_add functionality.

File: R/cnd_df.R [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+#'
+#' @param dat A data frame
+#' @param ... A set of logical conditions, e.g. `y & z, x | z` (`x`, `y`, `z`
+#'   would have to exist either as columns in `dat` or in the environment
+#'   `.env`). If multiple expressions are included, they are combined with the
+#'   `&` operator.
+#' @param .na Return value to be used when the conditions evaluate to `NA`.
+#' @param .env An optional environment to look for variables involved in logical
+#'   expression passed in `...`. A data frame or a list can also be passed that
+#'   will be coerced to an environment internally.
```
[the remark this answers is not in the document]
Comment (ramiromagno) [reply R/cnd_df.R:201]: If you're passing `dat2` explicitly you don't need to indicate the environment it resides in as it is being passed on. In that case it is really just a matter of renaming `.env` to `.dat2` --- note that I am adding again the dot on the parameter name because `condition_by()` accepts `...` and having a parameter name starting with dot signals that.

File: R/cnd_df.R [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+#' df_with_na <- data.frame(
+#'   x = c(1, 2, NA, 4, 5),
+#'   y = c(TRUE, FALSE, TRUE, FALSE, TRUE)
+#' )
+#' sdtm.oak:::eval_conditions(df_with_na, x > 2, .na = FALSE)
+#'
+#' # The environment where `eval_conditions()` is called is also inspected
+#' # when evaluating conditions in `...`.
+#' w <- 1
+#' sdtm.oak:::eval_conditions(df, x > w)
```
[the remark this answers is not in the document]
Comment (ramiromagno) [reply R/cnd_df.R:229]: > It will be helpful for the user to specify the dataframe and then specify the environment it resides in as two separate arguments.

The intention in having `.env` is not to indicate where the main data frame resides but to specify a namespace for looking for other variables that might be needed in the conditions supplied and that do not exist in the data frame `dat`. So you really should see `.env` as an object that provides a bag full of variables/vectors to be looked up for evaluation in the conditions given in `...` (if not part already of `dat`).

Also note that: `eval_conditions()` is currently an internal function, so it is not meant to be used by the user directly.

File: R/sdtm_join.R [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+#'
+#' [sdtm_join()] is a special join between a raw data set and a target data
+#' set. This function supports conditioned data frames.
+#'
+#' @param raw_dat The raw dataset: a dataframe or a conditioned data frame. Must
+#'   include the variables passed in `id_vars`.
+#' @param tgt_dat Target dataset: a data frame or a conditioned data frame to be
+#'   merged against `raw_dat` by the variables indicated in `id_vars`.
+#' @param id_vars Key variables to be used in the join between the raw dataset
+#'   (`raw_dat`) and the target data set (`tgt_dat`).
```
[the remark this answers is not in the document]
Comment (ramiromagno) [reply R/sdtm_join.R:11]: Again, `sdtm_join()` is an internal function. Nonetheless it has an `id_vars` parameter and therefore whichever function calls `sdtm_join()` can supply other than default id variables.

Explanation by why of an example: `assign_no_ct()` calls `sdtm_assign()` that calls `sdtm_join()`. `assign_no_ct()` already supports a parameter for indicating id variables: `id_vars`. So the user can:

- either fully specify `id_vars` by `assign_no_ct(id_vars = c("my_id1", "my_id2", etc.))`
- or use the default ones plus a few extra with  `assign_no_ct(id_vars = oak_id_vars(c("my_id1", "my_id2", etc.)))`.

So, to my understanding, your request is already implemented.

File: R/cnd_df.R (hunk 1 above)
Comment (ramiromagno) [reply R/cnd_df.R:23]: Sure.

**Comment by ramiromagno** [comment]
> Greate Job, @ramiromagno . I have added my comments. We can additional test cases and examples. Also, it will be good to change the way we compare two datasets, by adding an extra argument as suggested.

Thanks @rammprasad. I am not sure what was your take on relocating `tgt_dat` and make it a first argument. If you agree perhaps moving also `tgt_var` and make it the second argument would also make sense I reckon.

File: R/cnd_df.R (hunk 3 above)
[the remark this answers is not in the document]
Comment (rammprasad) [reply R/cnd_df.R:229]: ok, that makes it easier. We will just specify the .dat2 as the additional parameter.

File: R/cnd_df.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (rammprasad) [reply R/cnd_df.R:201]: `.dat2` works well. Thank you.

[the remark this answers is not in the document]
Comment (rammprasad) [reply R/cnd_df.R:201]: Also, let's change the order of the arguments so the user doesn't have to give additional parenthesis {} while using `condition_add`

```
assign_no_ct <-(
tgt_dat = NULL,
tgt_var,
raw_dat,
raw_var,
id_vars = oak_id_vars())

assign_ct <-(
tgt_dat = NULL,
tgt_var,
raw_dat,
raw_var,
ct_spec,
ct_clst,
id_vars = oak_id_vars())
```

But the downside is that the parameters with default values are not grouped together and they are not at the end. As you have specified, it will make the piping much better. Let's go for it.

**Comment by ramiromagno** [comment]
I thought we could circumvent the need for braces when using magrittr's pipe placeholder in nested calls if we moved `tgt_dat` to being the first argument but seemingly we cannot.

So I think we are left only with three options:

1. Move the `condition_add()` one level up and have it in the chain, i.e. not inline within the `assign_no_ct()`
2. Add braces around the `assign_no_ct()` call and use the placeholder where needed
3. Study the possibility of having another pipe operator whose behavior is equivalent to `%>% {...}`, but with the advantage of not needing the braces. I asked a question about this here: https://github.com/tidyverse/magrittr/issues/272.

In my opinion, the simplest and easiest for the user is option 1, i.e. moving `condition_add()` to an earlier position in the chain of commands.

``` r
library(sdtm.oak)
library(magrittr)
library(dplyr, warn.conflicts = FALSE)

study_ct <- read.csv(system.file("cm_domain/cm_sdtm_oak_ct.csv", package = "sdtm.oak"))
cm_raw <-  read.csv(system.file("cm_domain/cm_raw_data.csv", package = "sdtm.oak")) %>%
  #derive oak_id as the row number
  dplyr::mutate(
    oak_id = structure(seq_len(nrow(.))),
    patient_number = PATNUM,
    raw_source = "ConMed"
  ) %>%
  dplyr::select(oak_id_vars(), dplyr::everything())

# DOES NOT WORK (native pipe)
# assign_no_ct(raw_dat = cm_raw,
#              raw_var = "MDRAW",
#              tgt_var = "CMTRT")  |>
#   assign_no_ct(
#     tgt_dat = condition_add(_, CMTRT == "BABY ASPIRIN"),
#     tgt_var = "CMGRPID",
#     raw_dat = cm_raw,
#     raw_var = "MDNUM"
#   )

# DOES NOT WORK EITHER (magrittr's pipe)
assign_no_ct(raw_dat = cm_raw,
             raw_var = "MDRAW",
             tgt_var = "CMTRT")  %>%
  assign_no_ct(
    tgt_dat = condition_add(., CMTRT == "BABY ASPIRIN"),
    tgt_var = "CMGRPID",
    raw_dat = cm_raw,
    raw_var = "MDNUM"
  )
#> Error in `admiraldev::assert_character_vector()` at sdtm.oak/R/assign.R:191:3:
#> ! `id_vars` must be a character vector but is a data frame

# WORKS (native pipe)
assign_no_ct(raw_dat = cm_raw,
             raw_var = "MDRAW",
             tgt_var = "CMTRT")  |>
  condition_add(CMTRT == "BABY ASPIRIN") |>
  assign_no_ct(
    raw_dat = cm_raw,
    raw_var = "MDNUM",
    tgt_var = "CMGRPID"
  )
#> # A tibble: 14 × 5
#>    oak_id raw_source patient_number CMTRT                         CMGRPID
#>     <int> <chr>               <int> <chr>                           <int>
#>  1      1 ConMed                375 BABY ASPIRIN                        1
#>  2      2 ConMed                375 CORTISPORIN                        NA
#>  3      3 ConMed                376 ASPIRIN                            NA
#>  4      4 ConMed                377 DIPHENHYDRAMINE HCL                NA
#>  5      5 ConMed                377 PARCETEMOL                         NA
#>  6      6 ConMed                377 VOMIKIND                           NA
#>  7      7 ConMed                377 ZENFLOX OZ                         NA
#>  8      8 ConMed                378 AMITRYPTYLINE                      NA
#>  9      9 ConMed                378 BENADRYL                           NA
#> 10     10 ConMed                378 DIPHENHYDRAMINE HYDROCHLORIDE      NA
#> 11     11 ConMed                378 TETRACYCLINE                       NA
#> 12     12 ConMed                379 BENADRYL                           NA
#> 13     13 ConMed                379 SOMINEX                            NA
#> 14     14 ConMed                379 ZQUILL                             NA

# WORKS (maggritr's pipe)
assign_no_ct(raw_dat = cm_raw,
             raw_var = "MDRAW",
             tgt_var = "CMTRT")  %>%
  condition_add(CMTRT == "BABY ASPIRIN") |>
  assign_no_ct(
    raw_dat = cm_raw,
    raw_var = "MDNUM",
    tgt_var = "CMGRPID"
  )
#> # A tibble: 14 × 5
#>    oak_id raw_source patient_number CMTRT                         CMGRPID
#>     <int> <chr>               <int> <chr>                           <int>
#>  1      1 ConMed                375 BABY ASPIRIN                        1
#>  2      2 ConMed                375 CORTISPORIN                        NA
#>  3      3 ConMed                376 ASPIRIN                            NA
#>  4      4 ConMed                377 DIPHENHYDRAMINE HCL                NA
#>  5      5 ConMed                377 PARCETEMOL                         NA
#>  6      6 ConMed                377 VOMIKIND                           NA
#>  7      7 ConMed                377 ZENFLOX OZ                         NA
#>  8      8 ConMed                378 AMITRYPTYLINE                      NA
#>  9      9 ConMed                378 BENADRYL                           NA
#> 10     10 ConMed                378 DIPHENHYDRAMINE HYDROCHLORIDE      NA
#> 11     11 ConMed                378 TETRACYCLINE                       NA
#> 12     12 ConMed                379 BENADRYL                           NA
#> 13     13 ConMed                379 SOMINEX                            NA
#> 14     14 ConMed                379 ZQUILL                             NA

# DOES NOT WORK (native pipe)
# assign_no_ct(raw_dat = cm_raw,
#              raw_var = "MDRAW",
#              tgt_var = "CMTRT")  |>
#   {
#     assign_no_ct(
#       tgt_dat = condition_add(_, CMTRT == "BABY ASPIRIN"),
#       raw_dat = cm_raw,
#       raw_var = "MDNUM",
#       tgt_var = "CMGRPID"
#     )
#   }

# WORKS (maggritr's pipe)
assign_no_ct(raw_dat = cm_raw,
             raw_var = "MDRAW",
             tgt_var = "CMTRT")  %>%
  {
  assign_no_ct(
    tgt_dat = condition_add(., CMTRT == "BABY ASPIRIN"),
    raw_dat = cm_raw,
    raw_var = "MDNUM",
    tgt_var = "CMGRPID"
  )
  }
#> # A tibble: 14 × 5
#>    oak_id raw_source patient_number CMTRT                         CMGRPID
#>     <int> <chr>               <int> <chr>                           <int>
#>  1      1 ConMed                375 BABY ASPIRIN                        1
#>  2      2 ConMed                375 CORTISPORIN                        NA
#>  3      3 ConMed                376 ASPIRIN                            NA
#>  4      4 ConMed                377 DIPHENHYDRAMINE HCL                NA
#>  5      5 ConMed                377 PARCETEMOL                         NA
#>  6      6 ConMed                377 VOMIKIND                           NA
#>  7      7 ConMed                377 ZENFLOX OZ                         NA
#>  8      8 ConMed                378 AMITRYPTYLINE                      NA
#>  9      9 ConMed                378 BENADRYL                           NA
#> 10     10 ConMed                378 DIPHENHYDRAMINE HYDROCHLORIDE      NA
#> 11     11 ConMed                378 TETRACYCLINE                       NA
#> 12     12 ConMed                379 BENADRYL                           NA
#> 13     13 ConMed                379 SOMINEX                            NA
#> 14     14 ConMed                379 ZQUILL                             NA
```

File: R/cnd_df.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (ramiromagno) [reply R/cnd_df.R:201]: `.env` has been renamed to `.dat2` in `condition_add()`.

File: R/sdtm_join.R (hunk 4 above)
[the remark this answers is not in the document]
Comment (ramiromagno) [reply R/sdtm_join.R:11]: Made `oak_id_vars()` an exported function to make this possible.

File: R/cnd_df.R (hunk 1 above)
Comment (ramiromagno) [reply R/cnd_df.R]: Done.

**Comment by rammprasad** [comment]
Thank you, @ramiromagno. Regarding the condition_add options, let's stick with options 1 and 2. We will add examples for both scenarios.

1. Move the condition_add() one level up and have it in the chain, i.e. not inline within the assign_no_ct()
2. Add braces around the assign_no_ct() call and use the placeholder where needed

**Comment by rammprasad** [comment]
Let me know once it is ready for the final review.

**Comment by rammprasad** [comment]
Maybe we can add this as the third option. This looks sleeker than using {}

**Comment by ramiromagno** [comment]
@rammprasad and @edgar-manukyan:

I think we are pretty close to having the code for conditioned data frames near completion.

- A new pipe operator has been introduced `%.>%` check the docs for the details. This should allow to create chains of commands with less clutter, namely the usage of braces (`%>% {...}` can now be replaced simply with `%.>% ...`).
- I've added more documentation across functions and examples.
- I also added quite significant number of unit tests to most functions, but not all.
- A new vignette has been created introducing conditioned data frames. It is currently incomplete because it has only one usage example with `condition_add()`. We should add those cases Ram mentioned where conditioning involves either the raw data set, the target, both independently, and both interdependently. For the most complicated case, _both interdependently_ I reckon we will need to resort to using the `sdtm_join()` function explicitly by the user.

Take a look and give me your feedback!

**Comment by edgar-manukyan** [comment]
Simply brilliant @ramiromagno, thank you so much 🙏 🙏 🙏 for all your time and effort. I am sure the SDTM community is going to appreciate this. I also feel that admiral might grab your idea of conditioned data frames data as well 😉

**Review by edgar-manukyan — APPROVED** [review]

**Comment by rammprasad** [comment]
It looks good to me. Lets merge this to main, and I can take care of the documentation updates.

**Review by rammprasad — APPROVED** [review]

**Comment by ramiromagno** [comment]
@rammprasad and @edgar-manukyan : please do not merge yet as I am now doing styling and linting fixes.

File: R/assertions.R [hunk 5]
```diff
@@ -0,0 +1,16 @@
+# Surprisingly, admiraldev doesn't provide `assert_logical_vector`.
```
[the remark this answers is not in the document]
Comment (ramiromagno) [reply R/assertions.R:1]: Thanks @edgar-manukyan, good idea!

I've opened an issue first to see if they find it appropriate: https://github.com/pharmaverse/admiraldev/issues/455.

I'm happy to contribute with a PR if the authors agree.
